Foundation/FileManager.swift

```diff
@@ -2377,46 +2377,123 @@ extension FileManager {
+                }
+            }
+            let result = withURLCString(url: baseURL) { pszFrom -> BOOL? in
+                withURLCString(url: url) { pszTo in
+                    let fromAttrs = GetFileAttributesW(pszFrom)
+                    let toAttrs = GetFileAttributesW(pszTo)
+                    guard fromAttrs != INVALID_FILE_ATTRIBUTES, toAttrs != INVALID_FILE_ATTRIBUTES else {
+                        return FALSE
+                    }
+                    return PathRelativePathToW(relativePath.baseAddress, pszFrom, fromAttrs, pszTo, toAttrs)
+                }
+            }
+
+            guard result == TRUE, let (path, _) = String.decodeCString(relativePath.baseAddress, as: UTF16.self) else {
+                return nil
+            }
 #else
             let path = url.path.replacingOccurrences(of: baseURL.path+"/", with: "")
-            _currentItemPath = path
-            return path
 #endif
+            _currentItemPath = path
+            return _currentItemPath
         }
     }
 
-#if os(Windows)
     internal class NSURLDirectoryEnumerator : DirectoryEnumerator {
-        internal typealias ErrorHandler = /* @escaping */ (URL, Error) -> Bool
+#if os(Windows)
+        var _options : FileManager.DirectoryEnumerationOptions
+        var _errorHandler : ((URL, Error) -> Bool)?
+        var _stack: [URL]
+        var _current: URL?
+        var _rootDepth : Int
 
-        init(url: URL, options: FileManager.DirectoryEnumerationOptions, errorHandler: ErrorHandler?) {
+        init(url: URL, options: FileManager.DirectoryEnumerationOptions, errorHandler: (/* @escaping */ (URL, Error) -> Bool)?) {
+            _options = options
+            _errorHandler = errorHandler
+            _stack = [url]
+            _rootDepth = url.pathComponents.count
+        }
+
+        override func nextObject() -> Any? {
+            func contentsOfDir(directory: URL) -> [URL]? {
+                var ffd: WIN32_FIND_DATAW = WIN32_FIND_DATAW()
+                let dirPath = directory.absoluteString + "\\*"
+                let h: HANDLE = dirPath.withCString(encodedAs: UTF16.self) {
+                  FindFirstFileW($0, &ffd)
+                }
+                guard h != INVALID_HANDLE_VALUE else { return nil }
+                defer { FindClose(h) }
+
+                var files: [URL] = []
+                repeat {
+                    let fileArr = Array<WCHAR>(
+                      UnsafeBufferPointer(start: &ffd.cFileName.0,
+                                          count: MemoryLayout.size(ofValue: ffd.cFileName)))
+                    let file = String(decodingCString: fileArr, as: UTF16.self)
+                    if file != "."
+                        && file != ".."
+                        && (!_options.contains(.skipsHiddenFiles)
+                            || (ffd.dwFileAttributes & DWORD(FILE_ATTRIBUTE_HIDDEN) == 0)) {
+                        files.append(URL(fileURLWithPath: "\(directory.absoluteString)\\\(file)"))
```

Comment:
No, we have an internal helper `joinPath`, please use that rather than the `"\(directory.absoluteString)\\\(file)"`.  Also, FSR the `directory`.